kiwi_keg/generator.py

```diff
@@ -153,6 +155,94 @@ def create_custom_scripts(self, override: bool = False):
+                            overlay_dest_dir = os.path.join(self.dest_dir, profile_name)
+
+                        overlay_name = os.path.basename(overlay_dest_dir)
+                        for overlay_path in overlay_content.get('include'):
+                            overlay_full_path = os.path.join(
+                                self.image_definition.overlay_root,
+                                overlay_path
+                            )
+                            # loop all the file paths of overlay sub directory 'overlay_path'
+                            for name in KegUtils.get_all_files(overlay_full_path):
+                                rel_path = os.path.relpath(name, overlay_full_path)
+                                new_dir = os.path.dirname(rel_path)
+                                if new_dir:
+                                    new_dir = os.path.join(overlay_dest_dir, new_dir)
+                                    os.makedirs(new_dir, exist_ok=True)
+                                dest_file = os.path.join(overlay_dest_dir, rel_path)
+                                shutil.copy(name, dest_file)
+
+                        tarball_data[overlay_name] = {}
+                        tarball_data[overlay_name] = overlay_dest_dir
+
+                    if tar_overlays:
+                        for overlay_name, overlay_dest_dir in tarball_data.items():
+                            overlay_tarball_name = '{}.tar.gz'.format(overlay_name)
+                            self._create_tarball(
+                                os.path.join(self.dest_dir, overlay_tarball_name),
+                                overlay_dest_dir
+                            )
+                            shutil.rmtree(overlay_dest_dir)
+                            if overlay_name != 'root':
+                                self._update_config_kiwi(
+                                    overlay_tarball_name,
+                                    overlay_dest_dir
+                                )
+
+        if tar_overlays and not has_overlays:
+            log.warn(
+                'Attempt to create a tarball but not overlay paths were provided.'
+            )
+
+    @staticmethod
+    def _create_tarball(tarball_dir, dest_dir):
+        with tarfile.open(tarball_dir, 'w:gz') as tar:
+            for overlay_dir in os.scandir(dest_dir):
+                tar.add(
+                    overlay_dir.path,
+                    arcname=overlay_dir.name
+                )
+
+    def _update_config_kiwi(self, archive_name, dest_dir):
+        if os.path.exists(dest_dir) and os.path.exists(self.kiwi_description):
+            etree_parser = etree.XMLParser(remove_blank_text=True)
+            kiwi_xml = etree.parse(self.kiwi_description, parser=etree_parser)
+            kiwi_root = kiwi_xml.getroot()
+
+            image_element = kiwi_root.find('.//packages[@type="image"]')
+
+            archive_element = etree.SubElement(image_element, 'archive')
```

Comment:
But the archive element should not be created if we have `root.tar.gz`

Reply:
Generating the kiwi file first without the archive tags and then modified it isn't so nice, and it also requires knowledge about its structure that is defined elsewhere. That is why I suggested a two-stage approach. First, process the overlay configuration info and add the appropriate tags to the dictionary that can be used for schema generation. Second, produce the tarballs. Did you have a look at my previous comment and the sample code?

I think copying the files into a temp dir might be unnecessary. It should be possible to loop over all files and add them to the tarball directly using `TarFile.AddFile`. The archive name of the file it can be set in the `TarInfo`, and at that opportunity the owner and group can be set to `root` as well, which we really should do anyway. Again, the sample code in https://github.com/SUSE-Enceladus/keg/commit/393259f9fbe781c41c3f2bd84874bbd85f95f1e4 preprocesses the overlay information in a way that would make it easy to do that.

Reply:
> I think copying the files into a temp dir might be unnecessary. 

But it is not a temp dir, in an introductory conversation with @schaefi it was agreed the creation of `.tar.gz` files would be optional, and keg will always create an overlay tree structure (if `overlayfiles` is present) 

is this not the case ?

Reply:
> But the archive element should not be created if we have `root.tar.gz`

I understood different, if root, do not add root to archive, AFAIU now is not even necessary.

Reply:
Now I get it, the flag only declares whether the overlay should be tar'd or not, it will always be produced. So yes, if producing unpacked `root/` is an option we must have, then the option makes sense. I don't really think this is a work-flow that should be encouraged. You should update the recipe, run keg, then kiwi, not run keg, meddle with the overlay files, run kiwi, which would appear the prime reason to me why you wouldn't want it archived. But just my opinion.

However, this only works for root, doesn't it? I.e. named archives always need to be tarballs, right? The code doesn't create tarballs for overlays != root if the `tar_overlays` option is not set.

Reply:
> But the archive element should not be created if we have `root.tar.gz`

Moved the check for root inside the method.